chore: adjust env vars so that check-env-var-annotations passes

phil-davis opened the pull request:
## Description
1) chore: adjust documentation of check-env-var-annotations.
2) display line numbers in the output of check-env-var-annotatrions. This makes it easy to find the exact line that has the problem.
3) adjust logic for annotation checks so that it really passes when there are no problems.
4) add introductionVersion to all environment variable documentation.
5) add missing descriptions to environment variable documentation.

I added some "vacuous" description text in the cases where I didn't really know what to write. Suggestions welcome.

## Related Issue
- Fixes #8434

## How Has This Been Tested?
Local run of:
`make check-env-var-annotations` gives:
```
.make/check-env-var-annotations.sh
All introductionVersion annotations are valid
All description annotations are valid
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

phil-davis commented on ocis-pkg/shared/shared_types.go:
```diff
@@ -80,9 +80,9 @@ type Commons struct {
 	Reva              *Reva           `yaml:"reva"`
 	MachineAuthAPIKey string          `mask:"password" yaml:"machine_auth_api_key" env:"OCIS_MACHINE_AUTH_API_KEY" desc:"Machine auth API key used to validate internal requests necessary for the access to resources from other services." introductionVersion:"pre5.0"`
-	TransferSecret    string          `mask:"password" yaml:"transfer_secret,omitempty" env:"REVA_TRANSFER_SECRET" introductionVersion:"pre5.0"`
+	TransferSecret    string          `mask:"password" yaml:"transfer_secret,omitempty" env:"REVA_TRANSFER_SECRET" desc:"reva transfer secret" introductionVersion:"pre5.0"`
```
what is a good description to put here?

phil-davis commented on ocis-pkg/shared/shared_types.go:
```diff
@@ -80,9 +80,9 @@ type Commons struct {
+	TransferSecret    string          `mask:"password" yaml:"transfer_secret,omitempty" env:"REVA_TRANSFER_SECRET" desc:"reva transfer secret" introductionVersion:"pre5.0"`
 	SystemUserID      string          `yaml:"system_user_id" env:"OCIS_SYSTEM_USER_ID" desc:"ID of the oCIS storage-system system user. Admins need to set the ID for the storage-system system user in this config option which is then used to reference the user. Any reasonable long string is possible, preferably this would be an UUIDv4 format." introductionVersion:"pre5.0"`
-	SystemUserAPIKey  string          `mask:"password" yaml:"system_user_api_key" env:"SYSTEM_USER_API_KEY" introductionVersion:"pre5.0"`
+	SystemUserAPIKey  string          `mask:"password" yaml:"system_user_api_key" env:"SYSTEM_USER_API_KEY" desc:"system user API key" introductionVersion:"pre5.0"`
```
what is a good description to put here?

dragonchaser replied:
```suggestion
	SystemUserAPIKey  string          `mask:"password" yaml:"system_user_api_key" env:"OCIS_SYSTEM_USER_API_KEY" desc:"API key for the system user." introductionVersion:"pre5.0"`
```

phil-davis commented on services/frontend/pkg/config/config.go:
```diff
@@ -35,7 +35,7 @@ type Config struct {
 	EnableFederatedSharingOutgoing bool   `yaml:"enable_federated_sharing_outgoing" env:"FRONTEND_ENABLE_FEDERATED_SHARING_OUTGOING" desc:"Changing this value is NOT supported. Enables support for outgoing federated sharing for clients. The backend behaviour is not changed." introductionVersion:"pre5.0"`
 	SearchMinLength                int    `yaml:"search_min_length" env:"FRONTEND_SEARCH_MIN_LENGTH" desc:"Minimum number of characters to enter before a client should start a search for Share receivers. This setting can be used to customize the user experience if e.g too many results are displayed." introductionVersion:"pre5.0"`
-	Edition                        string `yaml:"edition" env:"OCIS_EDITION;FRONTEND_EDITION" introductionVersion:"pre5.0"`
+	Edition                        string `yaml:"edition" env:"OCIS_EDITION;FRONTEND_EDITION" desc:"Edition of oCIS" introductionVersion:"pre5.0"`
```
what is a good description to put here?

dragonchaser replied:
```suggestion
	Edition                        string `yaml:"edition" env:"OCIS_EDITION;FRONTEND_EDITION" desc:"Edition of oCIS. Used for branding pruposes." introductionVersion:"pre5.0"`
```

phil-davis commented on services/ocdav/pkg/config/config.go:
```diff
@@ -81,5 +81,5 @@ type Status struct {
 	ProductName    string
 	ProductVersion string
-	Edition        string `yaml:"edition" env:"OCIS_EDITION;OCDAV_EDITION" introductionVersion:"pre5.0"`
+	Edition        string `yaml:"edition" env:"OCIS_EDITION;OCDAV_EDITION" desc:"Edition of oCIS" introductionVersion:"pre5.0"`
```
what is a good description to put here?

dragonchaser replied:
```suggestion
	Edition        string `yaml:"edition" env:"OCIS_EDITION;OCDAV_EDITION" desc:"Edition of oCIS. Used for branding pruposes." introductionVersion:"pre5.0"`
```

phil-davis commented on services/web/pkg/config/config.go:
```diff
@@ -100,7 +100,7 @@ type ExternalApp struct {
 // ExternalAppConfig defines an external web app configuration.
 type ExternalAppConfig struct {
-	URL string `json:"url,omitempty" yaml:"url" env:"" introductionVersion:"pre5.0"`
+	URL string `json:"url,omitempty" yaml:"url" env:"" desc:"" introductionVersion:"pre5.0"`
```
This had an `env` section that is empty. 
I added an empty `desc` section to keep the checker happy.
Should I delete them both?

dragonchaser replied:
```suggestion
	URL string `json:"url,omitempty" yaml:"url"`
```

dragonchaser replied:
Only configurable via yaml.

phil-davis commented:
@kulmann ready for review.

I can add the check to CI once we get these changes merged. It passes locally.

phil-davis commented:
ToDo:

I put `pre5.0` for all the env vars. But some of them might be introduced in 5.0.
How can I find out which are introduced in 5.0?

(Then I can adjust the `introductionVersion` for those env vars)

dragonchaser commented:
@phil-davis they are listed in that table: https://github.com/owncloud/ocis/blob/cc18f5c372b7d3795167addb3c50574a95717941/docs/services/general-info/env-var-deltas/4.0.0-5.0.0-added.md

phil-davis commented:
> @phil-davis they are listed in that table: https://github.com/owncloud/ocis/blob/cc18f5c372b7d3795167addb3c50574a95717941/docs/services/general-info/env-var-deltas/4.0.0-5.0.0-added.md

done - actually most of the env vars are added in 5.0. Hopefully that is correct.

phil-davis commented:
Now I just need some suggestions for the env vars that needed descriptions added.
See my questions above.

dragonchaser commented:
> Now I just need some suggestions for the env vars that needed descriptions added. See my questions above.

Added patch suggestions above.

phil-davis commented:
I rebased, resolved the conflicts with recent changes in master, and added `introductionVersion` 5.0 to some new env vars.

`make check-env-var-annotations` passes locally for me.

Please review. Then maybe we merge this now (to reduce the ongoing conflicts happening). And I can work on a separate PR to add `make check-env-var-annotations` to CI.

micbar commented:
@phil-davis Thanks so far! This is hard to review. I need to check, but a lot of env variables got a `5.0` introduction version which have already been there from the beginning.

phil-davis commented:
> @phil-davis Thanks so far! This is hard to review. I need to check, but a lot of env variables got a `5.0` introduction version which have already been there from the beginning.

I used the list in https://github.com/owncloud/ocis/blob/cc18f5c372b7d3795167addb3c50574a95717941/docs/services/general-info/env-var-deltas/4.0.0-5.0.0-added.md

That list is "long", and claims that a lot of env vars have been added in 5.0. If that list is not correct, then where is the next place to look to try and find out what was added in 5.0?

dragonchaser commented:
> > @phil-davis Thanks so far! This is hard to review. I need to check, but a lot of env variables got a `5.0` introduction version which have already been there from the beginning.
> 
> I used the list in https://github.com/owncloud/ocis/blob/cc18f5c372b7d3795167addb3c50574a95717941/docs/services/general-info/env-var-deltas/4.0.0-5.0.0-added.md
> 
> That list is "long", and claims that a lot of env vars have been added in 5.0. If that list is not correct, then where is the next place to look to try and find out what was added in 5.0?

Oh there might be some confusion here, sorry I forgot to mention that. The List always states the full List of variables if one of them has been added. e.g. there is a global variable starting with `OCIS_` and a service related one starting with `<SERVICENAME>_` the global has most likely been there before. But since both can configure the same thing, both are mentioned.

phil-davis commented:
> But since both can configure the same thing, both are mentioned.

For lines that have both variables, what is required. For example:
```
Level  string `mapstructure:"level" env:"OCIS_LOG_LEVEL;ANTIVIRUS_LOG_LEVEL" desc:"The log level. Valid values are: 'panic', 'fatal', 'error', 'warn', 'info', 'debug', 'trace'." introductionVersion:"5.0"`
```

If `OCIS_LOG_LEVEL` was introduced pre5.0 and `ANTIVIRUS_LOG_LEVEL` is introduced in 5.0 then what do we set `introductionVersion` to?

And the same question if it happened the other way around?

(this is just an example to get the principle of this, I haven't looked in detail to see exactly when these were introduced)

dragonchaser commented:
@phil-davis I think this is an agreement thing. Introduction version is only valid for the non-global part OR the global it is the only var in the line.

dragonchaser commented:
You are right... We might have to rethink the annotation.... maybe structured text like `OCIS_LOG_LEVEL:pre5.0;AUTH_SERVICE_LOG_LEVEL:5.0` but I am not sure about this....

dragonchaser commented:
@phil-davis I just have talked to @kobergj, in the foreseeable future we will only have one variable per line (see #8259). So for now it is fine to just use the version of the latest introduced variable. We do **NOT** need to mess around with structured text :relieved:

phil-davis commented:
Rebased, resolved conflict, added ` introductionVersion:"5.0"` to new env var lines added in recent PRs.

phil-davis commented:
I have done what I think is "correct" according to the documented list of env vars changed/added in 5.0
It might be true that most env var declaration did change, if various common env vars have been added and apply in many places.

Please try to review. I think that "review" is going to mean just seeing that "something reasonable" has happened to all the modified lines. To check each env var against the documented list will take you just as long as it took me to do it!

phil-davis commented:
Rebased again. Please let me know if/what you want me to do with this.

phil-davis commented:
See #8574
